backend/api/src/http/endpoints/session/oauth.rs

```diff
@@ -0,0 +1,199 @@
+                settings.is_local(),
+                &jwks,
+                &code,
+                settings.login_token_valid_duration,
+                settings.remote_target(),
+                redirect_kind,
+            )
+            .await?
+        }
+
+        other => return Err(anyhow::anyhow!("Unsupported OAuth request kind: {:?}", other).into()),
+    };
+
+    Ok(HttpResponse::Created().cookie(cookie).json(response))
+}
+
+async fn handle_google_oauth(
+    db: &PgPool,
+    config: &GoogleOAuth,
+    token_secret: &[u8; 32],
+    local_insecure: bool,
+    jwks: &jwk::JwkVerifier,
+    code: &str,
+    login_token_valid_duration: Option<Duration>,
+    remote_target: RemoteTarget,
+    redirect_kind: OAuthUrlKind,
+) -> Result<(CreateSessionResponse, Cookie<'static>), error::OAuth> {
+    let redirect_url = google::oauth_url(remote_target, redirect_kind);
+
+    let tokens = google::convert_oauth_code(config, code, &redirect_url).await?;
+
+    let claims = jwks.verify_oauth(&tokens.id_token, 3).await?;
+
+    let mut txn = db.begin().await?;
+
+    let google_auth = sqlx::query!(
+        "select user_id from user_auth_google where google_id = $1",
+        &claims.google_id
+    )
+    .fetch_optional(&mut txn)
+    .await?;
+
+    let (user_id, mask) = match &google_auth {
+        Some(google_auth) => {
+            // make sure that the user either has a profile, or can only *create* one.
+            let check_profile = sqlx::query!(
+                r#"select exists(select 1 from user_profile where user_id = $1) as "exists!""#,
+                google_auth.user_id
+            )
+            .fetch_one(&mut txn)
+            .await?;
+
+            let mask = if check_profile.exists {
+                SessionMask::GENERAL
+            } else {
+                SessionMask::PUT_PROFILE | SessionMask::DELETE_ACCOUNT
+            };
+
+            (google_auth.user_id, mask)
+        }
+        None => {
+            if !claims.email_verified {
+                return Err(error::OAuth::Google(error::GoogleOAuth::UnverifiedEmail));
+            }
+
+            let id = sqlx::query!(r#"insert into "user" default values returning id"#)
+                .fetch_one(&mut txn)
+                .await?
+                .id;
+
+            sqlx::query!(
+                r"insert into user_auth_google (user_id, google_id) values ($1, $2)",
+                id,
+                &claims.google_id
+            )
+            .execute(&mut txn)
+            .await?;
+
+            // todo: handle duplicate email here (return a conflict error)
+
+            sqlx::query!(
+                "insert into user_email (user_id, email) values ($1, $2::text)",
+                id,
+                &claims.email
+            )
+            .execute(&mut txn)
+            .await?;
+            (id, SessionMask::PUT_PROFILE)
+        }
+    };
+
+    let login_ttl = login_token_valid_duration.unwrap_or(Duration::weeks(2));
```

Comment:
repeated issue: sometimes the token is only is only valid for an hour, but the cookie is valid for 2 weeks at least